File: sdk-typescript/packages/marketplace/src/marketplace.ts
/**
 * WAVE SDK - Marketplace API
 *
 * Template, plugin, and asset marketplace.
 */

import type { WaveClient, PaginationParams, PaginatedResponse, Timestamps } from '@wave-av/core';

export type ItemType = "template" | "plugin" | "graphic" | "transition" | "audio_effect" | "theme";
export type ItemStatus = "draft" | "review" | "published" | "rejected" | "deprecated";

export interface MarketplaceItem extends Timestamps {
  id: string;
  name: string;
  description: string;
  type: ItemType;
  status: ItemStatus;
  author_id: string;
  author_name: string;
  version: string;
  price_cents: number;
  downloads: number;
  rating: number;
  rating_count: number;
  preview_url?: string;
  thumbnail_url?: string;
  tags?: string[];
  category: string;
}

export interface InstalledItem {
  id: string;
  item_id: string;
  organization_id: string;
  version: string;
  installed_at: string;
  auto_update: boolean;
}
export interface Review extends Timestamps {
  id: string;
  item_id: string;
  user_id: string;
  rating: number;
  comment: string;
}

export interface ListItemsParams extends PaginationParams {
  type?: ItemType;
  category?: string;
  min_rating?: number;
  max_price?: number;
  search?: string;
  order_by?: string;
  order?: "asc" | "desc";
}
export interface PublishRequest {
  name: string;
  description: string;
  type: ItemType;
  price_cents?: number;
  tags?: string[];
  category: string;
  file_url: string;
  preview_url?: string;
  thumbnail_url?: string;
}

/**
 * Template, plugin, and asset marketplace for browsing, installing, and publishing.
 *
 * @example
 * ```typescript
 * const items = await wave.marketplace.list({ type: "template", category: "overlays" });
 * await wave.marketplace.install(items.data[0].id);
 * ```
 */
export class MarketplaceAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/marketplace";
  constructor(client: WaveClient) {
    this.client = client;
  }

  async list(params?: ListItemsParams): Promise<PaginatedResponse<MarketplaceItem>> {
    return this.client.get<PaginatedResponse<MarketplaceItem>>(this.basePath, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async get(itemId: string): Promise<MarketplaceItem> {
    return this.client.get<MarketplaceItem>(`${this.basePath}/${itemId}`);
  }
  async install(itemId: string): Promise<InstalledItem> {
    return this.client.post<InstalledItem>(`${this.basePath}/${itemId}/install`);
  }
  async uninstall(itemId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${itemId}/install`);
  }
  async listInstalled(params?: PaginationParams): Promise<PaginatedResponse<InstalledItem>> {
    return this.client.get<PaginatedResponse<InstalledItem>>(`${this.basePath}/installed`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async publish(request: PublishRequest): Promise<MarketplaceItem> {
    return this.client.post<MarketplaceItem>(this.basePath, request);
  }
  async update(itemId: string, updates: Partial<PublishRequest>): Promise<MarketplaceItem> {
    return this.client.patch<MarketplaceItem>(`${this.basePath}/${itemId}`, updates);
  }
  async deprecate(itemId: string): Promise<void> {
    await this.client.post(`${this.basePath}/${itemId}/deprecate`);
  }
  async getReviews(itemId: string, params?: PaginationParams): Promise<PaginatedResponse<Review>> {
    return this.client.get<PaginatedResponse<Review>>(`${this.basePath}/${itemId}/reviews`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async addReview(itemId: string, review: { rating: number; comment: string }): Promise<Review> {
    return this.client.post<Review>(`${this.basePath}/${itemId}/reviews`, review);
  }
  async search(
    query: string,
    params?: ListItemsParams,
  ): Promise<PaginatedResponse<MarketplaceItem>> {
    return this.client.get<PaginatedResponse<MarketplaceItem>>(`${this.basePath}/search`, {
      params: { q: query, ...params } as Record<string, string | number | boolean | undefined>,
    });
  }
}

export function createMarketplaceAPI(client: WaveClient): MarketplaceAPI {
  return new MarketplaceAPI(client);
}

File: sdk-python/wave/marketplace.py
"""WAVE SDK - Marketplace API. Templates, plugins, and asset marketplace."""
from __future__ import annotations
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class MarketplaceItem(BaseModel):
    id: str; name: str; description: str; type: str; status: str; author_name: str; version: str; price_cents: int = 0; downloads: int = 0; rating: float = 0; tags: list[str] | None = None; category: str; created_at: str; updated_at: str

class MarketplaceAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/marketplace"
    def list(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def get(self, item_id: str) -> MarketplaceItem: return MarketplaceItem(**self._client.get(f"{self._base}/{item_id}"))
    def install(self, item_id: str) -> dict: return self._client.post(f"{self._base}/{item_id}/install")
    def uninstall(self, item_id: str) -> None: self._client.delete(f"{self._base}/{item_id}/install")
    def list_installed(self, **params: Any) -> dict: return self._client.get(f"{self._base}/installed", params={k: v for k, v in params.items() if v is not None})
    def publish(self, name: str, description: str, type: str, category: str, file_url: str, **kwargs: Any) -> MarketplaceItem: return MarketplaceItem(**self._client.post(self._base, json={"name": name, "description": description, "type": type, "category": category, "file_url": file_url, **kwargs}))
    def search(self, query: str, **params: Any) -> dict: return self._client.get(f"{self._base}/search", params={"q": query, **{k: v for k, v in params.items() if v is not None}})
    def add_review(self, item_id: str, rating: int, comment: str) -> dict: return self._client.post(f"{self._base}/{item_id}/reviews", json={"rating": rating, "comment": comment})
